File: backend/app/models/sms.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, Enum, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.core.database import Base


class SmsDirection(str, enum.Enum):
    INBOUND = "inbound"
    OUTBOUND = "outbound"


class SmsStatus(str, enum.Enum):
    PENDING = "pending"
    SENT = "sent"
    FAILED = "failed"
    RECEIVED = "received"


class TaskStatus(str, enum.Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"


class SmsMessage(Base):
    __tablename__ = "sms_messages"

    id = Column(Integer, primary_key=True, index=True)
    modem_id = Column(Integer, ForeignKey("modems.id"), nullable=False)
    direction = Column(Enum(SmsDirection), nullable=False)
    phone_number = Column(String(30), nullable=False)
    content = Column(Text, nullable=False)
    status = Column(Enum(SmsStatus), default=SmsStatus.PENDING)
    error_message = Column(Text)
    sent_at = Column(DateTime)
    received_at = Column(DateTime)
    created_at = Column(DateTime, default=datetime.utcnow)

    mm_sms_index = Column(String(20))  # mmcli SMS index, used for inbound dedup
    modem = relationship("Modem", back_populates="sms_messages")
    scheduled_task_id = Column(Integer, ForeignKey("sms_scheduled_tasks.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)


class SmsTemplate(Base):
    __tablename__ = "sms_templates"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    content = Column(Text, nullable=False)
    variables = Column(JSON)  # list of variable names like ["name", "code"]
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class SmsScheduledTask(Base):
    __tablename__ = "sms_scheduled_tasks"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    modem_id = Column(Integer, ForeignKey("modems.id"), nullable=False)
    recipients = Column(JSON, nullable=False)   # list of phone numbers
    content = Column(Text, nullable=False)
    cron_expression = Column(String(100))       # e.g. "0 9 * * *"
    send_once_at = Column(DateTime)             # for one-time scheduled sends
    status = Column(Enum(TaskStatus), default=TaskStatus.ACTIVE)
    last_run_at = Column(DateTime)
    next_run_at = Column(DateTime)
    run_count = Column(Integer, default=0)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    modem = relationship("Modem", back_populates="scheduled_tasks")

File: backend-go/models/sms.go
package models

import "time"

// SMS direction / status / task status values.
const (
	SmsInbound  = "inbound"
	SmsOutbound = "outbound"

	SmsPending  = "pending"
	SmsSent     = "sent"
	SmsFailed   = "failed"
	SmsReceived = "received"

	TaskActive    = "active"
	TaskPaused    = "paused"
	TaskCompleted = "completed"
	TaskFailed    = "failed"
)

type SmsMessage struct {
	ID           uint       `gorm:"primaryKey" json:"id"`
	ModemID      uint       `gorm:"not null" json:"modem_id"`
	Direction    string     `gorm:"size:16;not null" json:"direction"`
	PhoneNumber  string     `gorm:"size:30;not null" json:"phone_number"`
	Content      string     `gorm:"type:text;not null" json:"content"`
	Status       string     `gorm:"size:16;default:pending" json:"status"`
	ErrorMessage *string    `gorm:"type:text" json:"error_message"`
	SentAt       *time.Time `json:"sent_at"`
	ReceivedAt   *time.Time `json:"received_at"`
	CreatedAt    time.Time  `json:"created_at"`

	MmSmsIndex      string `gorm:"column:mm_sms_index;size:20" json:"mm_sms_index"`
	ScheduledTaskID *uint  `json:"scheduled_task_id"`
	CreatedByID     *uint  `json:"created_by_id"`
}

func (SmsMessage) TableName() string { return "sms_messages" }

type SmsScheduledTask struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	Name           string     `gorm:"size:100;not null" json:"name"`
	ModemID        uint       `gorm:"not null" json:"modem_id"`
	Recipients     JSONList   `gorm:"type:json;not null" json:"recipients"`
	Content        string     `gorm:"type:text;not null" json:"content"`
	CronExpression *string    `gorm:"size:100" json:"cron_expression"`
	SendOnceAt     *time.Time `json:"send_once_at"`
	Status         string     `gorm:"size:16;default:active" json:"status"`
	LastRunAt      *time.Time `json:"last_run_at"`
	NextRunAt      *time.Time `json:"next_run_at"`
	RunCount       int        `gorm:"default:0" json:"run_count"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
	CreatedByID    *uint      `json:"created_by_id"`
}

func (SmsScheduledTask) TableName() string { return "sms_scheduled_tasks" }
